packages/dd-trace/src/latency_stats.js

```diff
@@ -0,0 +1,156 @@
+const os = require('os')
+const pkg = require('../../../package.json')
+const { decodePathwayContext } = require('../../datadog-plugin-kafkajs/src/hash')
+
+const { LogCollapsingLowestDenseDDSketch } = require('@datadog/sketches-js')
+
+const { LatencyStatsExporter } = require('./exporters/latency-stats')
+
+class AggStats {
+  constructor (aggKey) {
+    this.hash = aggKey.hash
+    this.parentHash = aggKey.parentHash
+    this.edgeTags = aggKey.edgeTags
+    this.edgeLatency = new LogCollapsingLowestDenseDDSketch(0.00775)
+    this.pathwayLatency = new LogCollapsingLowestDenseDDSketch(0.00775)
+  }
+
+  record (checkpoint) {
+    const edgeLatency = checkpoint.metrics.edge_latency
+    const pathwayLatency = checkpoint.metrics.pathway_latency
+    this.edgeLatency.accept(edgeLatency)
+    this.pathwayLatency.accept(pathwayLatency)
+  }
+
+  toJSON () {
+    return {
+      Hash: this.hash,
+      ParentHash: this.parentHash,
+      EdgeTags: this.edgeTags,
+      EdgeLatency: this.edgeLatency.toProto(), // TODO: custom proto encoding
+      PathwayLatency: this.pathwayLatency.toProto() // TODO: custom proto encoding
+    }
+  }
+}
+
+class AggKey {
+  constructor (checkpoint) {
+    this.hash = decodePathwayContext(checkpoint.metrics['dd-pathway-ctx'])[0]
+    this.parentHash = checkpoint.metrics.parent_hash
+    this.edgeTags = checkpoint.metrics.edge_tags
+  }
+
+  toString () {
+    return [
+      this.hash.toString(),
+      this.parentHash.toString()
+    ].join(',')
+  }
+}
+
+class SpanBuckets extends Map {
+  forCheckpoint (checkpoint) {
+    const aggKey = new AggKey(checkpoint)
+    const key = aggKey.toString()
+    // also include parentHash, edgeTags in ddsketch
+    if (!this.has(key)) {
+      this.set(key, new AggStats(aggKey)) // StatsPoint
+    }
+
+    return this.get(key)
+  }
+}
+
+class TimeBuckets extends Map {
+  forTime (time) {
+    if (!this.has(time)) {
+      this.set(time, new SpanBuckets())
+    }
+
+    return this.get(time)
+  }
+}
+
+class LatencyStatsProcessor {
+  constructor ({
+    dsmEnabled,
+    hostname,
+    port,
+    url,
+    env,
+    tags
+  } = {}) {
+    this.exporter = new LatencyStatsExporter({
+      hostname,
+      port,
+      tags,
+      url
+    })
+    this.bucketSizeNs = 1e10
+    this.buckets = new TimeBuckets()
+    this.hostname = os.hostname()
+    this.enabled = dsmEnabled
+    this.env = env
+    this.tags = tags || {}
+    this.sequence = 0
+
+    if (this.enabled) {
+      this.timer = setInterval(this.onInterval.bind(this), 1e4) // TODO: the right interval?
```

Comment:
`1e4` here might be better written as `10_000` as it conveys to the reader quickly what the value is, reading as ten thousands, or ten seconds when considering a second is a thousand milliseconds.